File: data-pipeline/scripts/fetch_members.py
"""Fetch members of the current Congress and their sponsored/cosponsored bills.

Reads CONGRESS_API_KEY from the environment, runs in two sequential phases:

  Phase 1 (fast, ~7 min): walk /member/congress/{c} + per-member detail and
  publish docs/data/members_{NNN}.json. The Reps tab works as soon as this
  lands, so this phase is unconditional — it always runs to completion and
  the time budget does not gate it.

  Phase 2 (slow, paginated): walk each member's sponsored-legislation and
  cosponsored-legislation endpoints and write
  docs/data/members/{bioguideId}_{kind}.json. Senior senators have thousands
  of cosponsorships; this phase is gated by --time-budget-minutes and is
  self-resumable via per-member file existence (skips members whose two
  legislation files already exist).

The output `id` field on each bill matches the {type}{number}-{congress}
shape used by the existing bills manifest, so the Android app can do an
O(1) "is this bill in our cache?" lookup.
"""

from __future__ import annotations

import argparse
import os
import sys
import time
from datetime import datetime, timezone
from typing import Any, Iterable

from _common import (
    CongressClient,
    ErrorCollector,
    LIST_PAGE_LIMIT,
    current_congress,
    load_members_index,
    member_legislation_path,
    now_iso,
    parse_member_legislation_item,
    parse_member_summary,
    save_member_legislation,
    save_members_index,
)

# Soft caps to bound a runaway pagination loop. Current Congress has ~535
# voting members; the top cosponsorship counts in recent Congresses are
# under 5,000 per (member, kind). Both ceilings have ~5x headroom — bump
# them only if real-world runs hit the limit.
LIST_PAGES_MAX = 10
LEG_PAGES_MAX = 5  # 5 pages × 250 = 1250 items max per (member, kind).
                   # Senior senators have thousands of cosponsorships; capping
                   # at 1250 keeps runtime bounded without hurting app UX
                   # (we display at most a couple hundred at a time).


def list_members(client: CongressClient, congress: int) -> Iterable[dict[str, Any]]:
    """Yield member summaries from /member/congress/{congress}."""
    for page in range(LIST_PAGES_MAX):
        offset = page * LIST_PAGE_LIMIT
        body = client.get(
            f"/member/congress/{congress}",
            limit=LIST_PAGE_LIMIT,
            offset=offset,
            currentMember="true",
        )
        members = body.get("members") or []
        if not members:
            return
        yield from members
        # Heuristic termination: Congress.gov returns a short final page when
        # the list is exhausted. The `currentMember=true` filter doesn't
        # currently break this in observed responses, but verify if behavior
        # changes in a future API version.
        if len(members) < LIST_PAGE_LIMIT:
            return


def fetch_legislation(
    client: CongressClient, bioguide_id: str, kind: str
) -> list[dict[str, Any]]:
    """Fetch all pages of sponsored or cosponsored legislation for a member."""
    if kind == "sponsored":
        endpoint_path = f"/member/{bioguide_id}/sponsored-legislation"
        body_key = "sponsoredLegislation"
    elif kind == "cosponsored":
        endpoint_path = f"/member/{bioguide_id}/cosponsored-legislation"
        body_key = "cosponsoredLegislation"
    else:
        raise ValueError(f"unknown kind: {kind!r}")

    items: list[dict[str, Any]] = []
    for page in range(LEG_PAGES_MAX):
        offset = page * LIST_PAGE_LIMIT
        body = client.get(endpoint_path, limit=LIST_PAGE_LIMIT, offset=offset)
        page_items = body.get(body_key) or []
        if not page_items:
            break
        items.extend(page_items)
        if len(page_items) < LIST_PAGE_LIMIT:
            break
    return items


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--time-budget-minutes", type=int, default=300,
        help="Soft time budget for Phase 2 (default 300 = 5 hours). Phase 1 "
             "always runs to completion (it's fast). When the budget elapses "
             "during Phase 2, the script writes outputs and exits 0. GitHub "
             "Actions has a 6h hard cap; the default leaves 1h headroom.",
    )
    parser.add_argument(
        "--phase1-only", action="store_true",
        help="Run only Phase 1 (publish the members index) and exit. Use this "
             "when the workflow commits Phase 1 output before invoking the "
             "script a second time for Phase 2 — that pattern keeps the index "
             "on origin within ~10 minutes regardless of Phase 2 duration.",
    )
    parser.add_argument(
        "--phase2-only", action="store_true",
        help="Run only Phase 2 (backfill sponsored/cosponsored legislation). "
             "Phase 2 reads the existing index from disk, so a prior Phase 1 "
             "(or a committed members_NNN.json) must be in place.",
    )
    args = parser.parse_args(argv)
    if args.phase1_only and args.phase2_only:
        print("--phase1-only and --phase2-only are mutually exclusive.", file=sys.stderr)
        return 2
    start_time = time.time()
    deadline = start_time + args.time_budget_minutes * 60

    api_key = os.environ.get("CONGRESS_API_KEY")
    if not api_key:
        print("CONGRESS_API_KEY is not set in the environment.", file=sys.stderr)
        return 2

    congress = current_congress(datetime.now(timezone.utc))
    client = CongressClient(api_key)

    run_phase1 = not args.phase2_only
    run_phase2 = not args.phase1_only

    existing_index = load_members_index(congress) or {"members": []}
    existing_by_bid: dict[str, dict[str, Any]] = {
        m["bioguide_id"]: m for m in existing_index.get("members", [])
    }
    members_out: list[dict[str, Any]] = list(existing_index.get("members", []))

    phase1_errors = ErrorCollector()
    phase2_errors = ErrorCollector()

    # ---------- Phase 1: fast — member roster + detail only ----------
    if not run_phase1:
        print(
            f"Skipping Phase 1; using cached index with {len(members_out)} members."
        )
        if not members_out:
            print(
                "ERROR: --phase2-only requested but no existing index was found.",
                file=sys.stderr,
            )
            return 2
    else:
        print(f"Phase 1: fetching member index for the {congress}th Congress")
        members_out = []
        for summary in list_members(client, congress):
            bioguide_id = summary.get("bioguideId") or ""
            if not bioguide_id:
                continue
            try:
                detail_body = client.get(f"/member/{bioguide_id}")
                detail = detail_body.get("member") or {}
                # Detail body wins on every overlapping key; list summary supplies
                # fields the detail endpoint sometimes omits (e.g. depiction, terms).
                # The order matters — Congress.gov detail responses are richer for
                # most fields but can drop these specific ones intermittently.
                merged: dict[str, Any] = {**summary, **detail}
                parsed = parse_member_summary(merged)
            except Exception as exc:  # noqa: BLE001
                # Fall back to cached record from previous run if we have one;
                # otherwise drop the member silently. The failure goes into
                # the collector either way so the end-of-run summary surfaces
                # systematic upstream issues regardless of cache hit/miss.
                if bioguide_id in existing_by_bid:
                    parsed = existing_by_bid[bioguide_id]
                    phase1_errors.record("member_detail_reused_cache", bioguide_id, exc)
                else:
                    phase1_errors.record("member_detail_dropped", bioguide_id, exc)
                    continue
            members_out.append(parsed)
            print(f"  + {bioguide_id} {parsed['name']} ({parsed['party']}-{parsed['state']})")

        # Publish the index. The Reps tab works as soon as this lands.
        save_members_index(congress, {
            "congress": congress,
            "generated_at": now_iso(),
            "members": members_out,
        })
        print(f"Phase 1 done: index has {len(members_out)} members")
        phase1_errors.print_summary(label="Phase 1")

    if not run_phase2:
        print("Skipping Phase 2 (--phase1-only).")
        return 0

    # ---------- Phase 2: slow — per-member legislation backfill ----------
    print("\nPhase 2: backfilling sponsored/cosponsored legislation")
    backfilled = 0
    skipped_cached = 0
    for parsed in members_out:
        if time.time() > deadline:
            print(
                f"\n[time-budget] Stopping Phase 2 early; processed "
                f"{backfilled} new + skipped {skipped_cached} cached. "
                f"Re-run to continue.",
                file=sys.stderr,
            )
            break
        bioguide_id = parsed["bioguide_id"]
        if (
            member_legislation_path(bioguide_id, "sponsored").exists()
            and member_legislation_path(bioguide_id, "cosponsored").exists()
        ):
            skipped_cached += 1
            continue
        for kind in ("sponsored", "cosponsored"):
            try:
                raw_items = fetch_legislation(client, bioguide_id, kind)
            except Exception as exc:  # noqa: BLE001
                phase2_errors.record(f"{kind}_legislation", bioguide_id, exc)
                raw_items = []
            bills = [parse_member_legislation_item(it) for it in raw_items]
            save_member_legislation(bioguide_id, kind, {
                "bioguide_id": bioguide_id,
                "congress": congress,
                "kind": kind,
                "generated_at": now_iso(),
                "bills": bills,
            })
        backfilled += 1
        print(f"  + {bioguide_id}: legislation backfilled")

    print(
        f"\nPhase 2 done: backfilled {backfilled} members, "
        f"skipped {skipped_cached} already-cached. "
        f"Index has {len(members_out)} total."
    )
    phase2_errors.print_summary(label="Phase 2")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: pipeline/shared/src/commonMain/kotlin/com/informedcitizen/pipeline/fetch/FetchMembers.kt
package com.informedcitizen.pipeline.fetch

import com.informedcitizen.pipeline.ErrorCollector
import com.informedcitizen.pipeline.http.CongressClient
import com.informedcitizen.pipeline.http.LegislatorsClient
import com.informedcitizen.pipeline.model.Member
import com.informedcitizen.pipeline.model.MemberLegislationItem
import com.informedcitizen.pipeline.model.SocialHandle
import kotlinx.coroutines.CancellationException
import kotlinx.datetime.Clock
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject

/** Page size for `/member/congress/{c}` and per-member legislation list endpoints. */
const val MEMBER_LIST_PAGE_LIMIT: Int = 250

/** Mirrors Python `fetch_members.LIST_PAGES_MAX = 10` (10 × 250 = 2500 caps the roster walk). */
const val MEMBER_LIST_PAGES_MAX: Int = 10

/**
 * Mirrors Python `fetch_members.LEG_PAGES_MAX = 5` (5 × 250 = 1250 caps
 * each `(member, kind)` walk). Senior senators have thousands of
 * cosponsorships; the cap keeps runtime bounded — the app shows at most
 * a couple hundred at a time.
 */
const val MEMBER_LEG_PAGES_MAX: Int = 5

/** Per-phase tallies returned by [fetchMembers]. */
data class FetchMembersResult(
    val congress: Int,
    val ranPhase1: Boolean,
    val ranPhase2: Boolean,
    val phase1MemberCount: Int,
    val phase2Backfilled: Int,
    val phase2SkippedCached: Int,
    val phase2TimedOut: Boolean,
    val contactInfoLoaded: Boolean,
    val contactInfoSize: Int,
    val socialsLoaded: Boolean,
    val socialsSize: Int,
)

/**
 * Streaming progress hooks. Defaulted to no-ops so the orchestrator
 * stays testable without log noise; the CLI wraps each with a `println`.
 */
data class FetchMembersProgress(
    val onPhase1Start: (congress: Int) -> Unit = {},
    val onPhase1Member: (member: Member) -> Unit = {},
    val onPhase1Done: (count: Int) -> Unit = {},
    val onPhase2Start: () -> Unit = {},
    val onPhase2MemberDone: (bioguideId: String) -> Unit = {},
    val onPhase2TimeBudgetExceeded: (processed: Int, skipped: Int) -> Unit = { _, _ -> },
    val onPhase2Done: (backfilled: Int, skipped: Int, total: Int) -> Unit = { _, _, _ -> },
    val onContactInfoLoaded: (size: Int, withForm: Int, withSite: Int) -> Unit = { _, _, _ -> },
    val onContactInfoFailed: (message: String) -> Unit = {},
    val onSocialsLoaded: (size: Int, totalHandles: Int) -> Unit = { _, _ -> },
    val onSocialsFailed: (message: String) -> Unit = {},
    val onStateWarning: (message: String) -> Unit = {},
)

/**
 * Full fetch-members orchestrator. Mirrors Python
 * `fetch_members.main`:
 *
 *  - Phase 1 (unconditional unless `runPhase1=false`): walk
 *    `/member/congress/{c}` + per-member detail, merge with the
 *    contact-info side index, persist `members_NNN.json`.
 *  - Phase 2 (skipped if `runPhase2=false`): walk each member's
 *    sponsored / cosponsored legislation. Per-member skip when both
 *    `members/{bid}_{kind}.json` files already exist. Time-budget gate
 *    surfaces via `onPhase2TimeBudgetExceeded` and the function returns
 *    normally (mirrors Python's exit-0 behavior).
 *
 * Phase 1 always runs to completion when enabled — the Reps tab in the
 * Android app depends on the index landing, and it's fast (~7 min for
 * 535 members), so the time budget only gates Phase 2.
 *
 * The `clockOverride` parameter is the deadline anchor for the
 * time-budget check (defaults to wall clock); tests inject a fake
 * [Clock] to drive deterministic budget-exhaustion behavior.
 */
suspend fun fetchMembers(
    congressClient: CongressClient,
    legislatorsClient: LegislatorsClient,
    congress: Int,
    nowIso: String,
    indexStore: FileMembersIndexStore,
    legislationStore: FileMemberLegislationStore,
    errors: ErrorCollector,
    runPhase1: Boolean = true,
    runPhase2: Boolean = true,
    timeBudgetMillis: Long = 300L * 60L * 1000L,
    progress: FetchMembersProgress = FetchMembersProgress(),
    clockOverride: Clock = Clock.System,
    maxListPages: Int = MEMBER_LIST_PAGES_MAX,
    maxLegPages: Int = MEMBER_LEG_PAGES_MAX,
): FetchMembersResult {
    val startMillis = clockOverride.now().toEpochMilliseconds()
    val deadlineMillis = startMillis + timeBudgetMillis

    val existingIndex = indexStore.load(congress)
    val existingByBid: Map<String, Member> =
        existingIndex?.members.orEmpty().associateBy { it.bioguideId }
    var membersOut: List<Member> = existingIndex?.members.orEmpty()

    var contactInfo: Map<String, LegislatorContactInfo> = emptyMap()
    var contactInfoLoaded = false
    if (runPhase1) {
        try {
            val text = legislatorsClient.fetchCurrent()
            contactInfo = parseContactInfoJson(text)
            contactInfoLoaded = true
            val withForm = contactInfo.values.count { it.contactForm != null }
            val withSite = contactInfo.values.count { it.website != null }
            progress.onContactInfoLoaded(contactInfo.size, withForm, withSite)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            // Non-fatal: members still publish with both fields null.
            errors.record(
                kind = "contact_info_index",
                identifier = "unitedstates",
                errorClass = e::class.simpleName ?: "Throwable",
                message = e.message ?: e.toString(),
            )
            progress.onContactInfoFailed(e.message ?: e.toString())
        }
    }

    var socialsIndex: Map<String, List<SocialHandle>> = emptyMap()
    var socialsLoaded = false
    if (runPhase1) {
        try {
            val socialsText = legislatorsClient.fetchSocialMedia()
            socialsIndex = parseSocialsJson(socialsText)
            socialsLoaded = true
            val totalHandles = socialsIndex.values.sumOf { it.size }
            progress.onSocialsLoaded(socialsIndex.size, totalHandles)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            // Non-fatal: members still publish with socials=[].
            errors.record(
                kind = "socials_index",
                identifier = "unitedstates",
                errorClass = e::class.simpleName ?: "Throwable",
                message = e.message ?: e.toString(),
            )
            progress.onSocialsFailed(e.message ?: e.toString())
        }
    }

    if (runPhase1) {
        progress.onPhase1Start(congress)
        val collected = mutableListOf<Member>()
        for (summary in listMembers(congressClient, congress, maxListPages)) {
            val bioguideId = summary.stringField("bioguideId").orEmpty()
            if (bioguideId.isEmpty()) continue
            val parsed: Member? = try {
                val detailBody = congressClient.get("/member/$bioguideId")
                val detail = detailBody.jsonObjectField("member") ?: JsonObject(emptyMap())
                // Detail wins on every overlapping key; list summary supplies
                // fields the detail endpoint sometimes omits (e.g. depiction,
                // terms). Order matters — detail responses are richer for most
                // fields but can intermittently drop these specific ones.
                val merged = JsonObject(summary.toMutableMap().apply { putAll(detail) })
                parseMemberSummary(merged, progress.onStateWarning)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                val cached = existingByBid[bioguideId]
                if (cached != null) {
                    errors.record(
                        kind = "member_detail_reused_cache",
                        identifier = bioguideId,
                        errorClass = e::class.simpleName ?: "Throwable",
                        message = e.message ?: e.toString(),
                    )
                    cached
                } else {
                    errors.record(
                        kind = "member_detail_dropped",
                        identifier = bioguideId,
                        errorClass = e::class.simpleName ?: "Throwable",
                        message = e.message ?: e.toString(),
                    )
                    null
                }
            }
            if (parsed == null) continue
            val info = contactInfo[bioguideId]
            val withContact = parsed.copy(
                contactForm = info?.contactForm,
                website = info?.website,
                socials = socialsIndex[bioguideId].orEmpty(),
            )
            collected += withContact
            progress.onPhase1Member(withContact)
        }
        indexStore.save(congress, collected, nowIso)
        membersOut = collected
        progress.onPhase1Done(collected.size)
    } else {
        // --phase2-only: must have an existing index to walk.
        if (membersOut.isEmpty()) {
            errors.record(
                kind = "phase2_without_index",
                identifier = "members_$congress",
                errorClass = "IllegalState",
                message = "Phase 2 requested but no existing members index was found.",
            )
            return FetchMembersResult(
                congress = congress,
                ranPhase1 = false,
                ranPhase2 = false,
                phase1MemberCount = 0,
                phase2Backfilled = 0,
                phase2SkippedCached = 0,
                phase2TimedOut = false,
                contactInfoLoaded = false,
                contactInfoSize = 0,
                socialsLoaded = false,
                socialsSize = 0,
            )
        }
    }

    if (!runPhase2) {
        return FetchMembersResult(
            congress = congress,
            ranPhase1 = runPhase1,
            ranPhase2 = false,
            phase1MemberCount = membersOut.size,
            phase2Backfilled = 0,
            phase2SkippedCached = 0,
            phase2TimedOut = false,
            contactInfoLoaded = contactInfoLoaded,
            contactInfoSize = contactInfo.size,
            socialsLoaded = socialsLoaded,
            socialsSize = socialsIndex.size,
        )
    }

    progress.onPhase2Start()
    var backfilled = 0
    var skippedCached = 0
    var timedOut = false
    for (parsed in membersOut) {
        if (clockOverride.now().toEpochMilliseconds() > deadlineMillis) {
            timedOut = true
            progress.onPhase2TimeBudgetExceeded(backfilled, skippedCached)
            break
        }
        val bid = parsed.bioguideId
        if (legislationStore.exists(bid, "sponsored") &&
            legislationStore.exists(bid, "cosponsored")
        ) {
            skippedCached++
            continue
        }
        for (kind in listOf("sponsored", "cosponsored")) {
            val rawItems: List<JsonObject> = try {
                fetchLegislation(congressClient, bid, kind, maxLegPages)
            } catch (e: CancellationException) {
                throw e
            } catch (e: Throwable) {
                errors.record(
                    kind = "${kind}_legislation",
                    identifier = bid,
                    errorClass = e::class.simpleName ?: "Throwable",
                    message = e.message ?: e.toString(),
                )
                emptyList()
            }
            val bills: List<MemberLegislationItem> = rawItems.map(::parseMemberLegislationItem)
            legislationStore.save(bid, kind, congress, bills, nowIso)
        }
        backfilled++
        progress.onPhase2MemberDone(bid)
    }
    progress.onPhase2Done(backfilled, skippedCached, membersOut.size)

    return FetchMembersResult(
        congress = congress,
        ranPhase1 = runPhase1,
        ranPhase2 = true,
        phase1MemberCount = membersOut.size,
        phase2Backfilled = backfilled,
        phase2SkippedCached = skippedCached,
        phase2TimedOut = timedOut,
        contactInfoLoaded = contactInfoLoaded,
        contactInfoSize = contactInfo.size,
        socialsLoaded = socialsLoaded,
        socialsSize = socialsIndex.size,
    )
}

/**
 * Paginated walk of `/member/congress/{congress}` with `currentMember=true`.
 * Mirrors Python `fetch_members.list_members`. Terminates when a page
 * comes back shorter than the limit (Congress.gov returns a short
 * final page when the list is exhausted).
 */
internal suspend fun listMembers(
    client: CongressClient,
    congress: Int,
    maxPages: Int,
): List<JsonObject> {
    val collected = mutableListOf<JsonObject>()
    for (page in 0 until maxPages) {
        val offset = page * MEMBER_LIST_PAGE_LIMIT
        val body = client.get(
            path = "/member/congress/$congress",
            params = mapOf(
                "limit" to MEMBER_LIST_PAGE_LIMIT.toString(),
                "offset" to offset.toString(),
                "currentMember" to "true",
            ),
        )
        val members = (body["members"] as? JsonArray)?.mapNotNull { it as? JsonObject }.orEmpty()
        if (members.isEmpty()) return collected
        collected += members
        if (members.size < MEMBER_LIST_PAGE_LIMIT) return collected
    }
    return collected
}

/**
 * Fetch all pages of sponsored / cosponsored legislation for [bioguideId].
 * Mirrors Python `fetch_members.fetch_legislation`.
 */
internal suspend fun fetchLegislation(
    client: CongressClient,
    bioguideId: String,
    kind: String,
    maxPages: Int,
): List<JsonObject> {
    val (endpoint, bodyKey) = when (kind) {
        "sponsored" -> "/member/$bioguideId/sponsored-legislation" to "sponsoredLegislation"
        "cosponsored" -> "/member/$bioguideId/cosponsored-legislation" to "cosponsoredLegislation"
        else -> error("unknown kind: '$kind'")
    }
    val items = mutableListOf<JsonObject>()
    for (page in 0 until maxPages) {
        val offset = page * MEMBER_LIST_PAGE_LIMIT
        val body = client.get(
            path = endpoint,
            params = mapOf(
                "limit" to MEMBER_LIST_PAGE_LIMIT.toString(),
                "offset" to offset.toString(),
            ),
        )
        val pageItems = (body[bodyKey] as? JsonArray)?.mapNotNull { it as? JsonObject }.orEmpty()
        if (pageItems.isEmpty()) break
        items += pageItems
        if (pageItems.size < MEMBER_LIST_PAGE_LIMIT) break
    }
    return items
}
